### PR TITLE
Add seed argument to space's sample method

### LINKED ISSUE #560: Seed argument on space's random sample method
**Describe the feature you'd like**
I would like to have set seed functionality in `Box` and `DiscreteSearchSpace`'s `.sample()` method 

**Is your feature request related to a problem? Please describe.**
This will help reproducibility for a research purposes

**Describe alternatives you've considered**
Simply add seed argument to the `.sample()` method  and pass it to the `tf.random` function that generate the random points

### PR DESCRIPTION
**Pull request by satrialoka**
Add seed argument for sample() method in space
this will close #560

### CODE REVIEW

**Comment by satrialoka** [comment]
I add `seed: Optional[int]` argument in `sample` method inside `SearchSpace` abstract base class. This assume that we always want seed option in `sample`. Since the other sample (non random based)  method that possibly does not need seed arg, can have different name. for example `sample_sobol`, `sample_lhd` etc

File: tests/unit/test_space.py [hunk 1]
```diff
@@ -167,6 +168,14 @@ def test_discrete_search_space_sampling(num_samples: int) -> None:
     assert len(samples) == num_samples
 
 
+@pytest.mark.parametrize("seed", [1, 42, 123])
```
Comment (hstojic) [inline tests/unit/test_space.py:171]: can you please add tests that make sure we get different samples when seed has not been set

File: trieste/space.py [hunk 2]
```diff
@@ -35,9 +35,10 @@ class SearchSpace(ABC):
     """
 
     @abstractmethod
-    def sample(self, num_samples: int) -> TensorType:
+    def sample(self, num_samples: int, seed: Optional[int] = None) -> TensorType:
         """
         :param num_samples: The number of points to sample from this search space.
+        :param seed: Random seed for reproducibility
```
Comment (hstojic) [inline trieste/space.py]:
```suggestion
        :param seed: Random seed for reproducibility.
```

File: trieste/space.py [hunk 3]
```diff
@@ -176,17 +177,21 @@ def __contains__(self, value: TensorType) -> bool | TensorType:
         tf.debugging.assert_shapes([(value, self.points.shape[1:])])
         return tf.reduce_any(tf.reduce_all(value == self._points, axis=1))
 
-    def sample(self, num_samples: int) -> TensorType:
+    def sample(self, num_samples: int, seed: Optional[int] = None) -> TensorType:
         """
         :param num_samples: The number of points to sample from this search space.
+        :param seed: Random seed for reproducibility
```
Comment (hstojic) [inline trieste/space.py]:
```suggestion
        :param seed: Random seed for reproducibility.
```

File: trieste/space.py [hunk 4]
```diff
@@ -324,20 +329,26 @@ def __contains__(self, value: TensorType) -> bool | TensorType:
 
         return tf.reduce_all(value >= self._lower) and tf.reduce_all(value <= self._upper)
 
-    def sample(self, num_samples: int) -> TensorType:
+    def sample(self, num_samples: int, seed: Optional[int] = None) -> TensorType:
         """
         Sample randomly from the space.
 
         :param num_samples: The number of points to sample from this search space.
+        :param seed: Random seed for reproducibility
```
Comment (hstojic) [inline trieste/space.py]:
```suggestion
        :param seed: Random seed for reproducibility.
```

**Review by hstojic — COMMENTED** [review]
looks good, some minor comments

File: tests/unit/test_space.py (hunk 1 above)
Comment (satrialoka) [reply tests/unit/test_space.py:171]: sure

**Review by hstojic — APPROVED** [review]
looks good now, thanks :)

**Comment by hstojic** [comment]
ready to merge?

**Comment by satrialoka** [comment]
Yes please!
